### ISSUE TITLE
 Implement a converter from NGE operations  to OSRD timetable update - Part 1

### ISSUE THREAD

**Issue by louisgreiner** [OpenRailAssociation/osrd#7539]
4th step of meta #7535 

## Acceptance criteria 

- [x] A timetable change implies an update of the NGE model
- [x] An action on trainruns* in NGE implies an update of the timetable
  - [x] Create trainrun
    - [x] Create valid trainrun with valid path => train schedule creation
  - [x] Update trainrun
    - [x] Update trainrun path (shorten and lengthen steps or only start, intermediary or end step) => train schedule updated 
  - [x] Delete trainrun
    - [x] Delete trainrun => delete train schedule

- [x] Information converted from NGE to OSRD : 
  - [x] Name
  - [x] Path

*a `trainrun` in NGE is the equivalent of a `train schedule` in OSRD (even tho it is not semantically 100% correct)

## Old acceptance criteria (when NGE was supposed to take the output of OSRD)

- [x] A timetable change implies an update of the NGE model
- [x] An action on trainruns* in NGE implies an update of the timetable
  - [ ] Create trainrun
    - [x] Create valid trainrun with valid path => train schedule creation (simulation not working, rolling stock must be entered in OSRD first) => once the rollingstock is given and simulation done => trainrun updated (running times, same path)
    - [ ] Create valid trainrun with incomplete path (should pass by intermediary stations) => train schedule creation (simulation not working, rolling stock must be entered in OSRD first) => once the rollingstock is given and simulation done => trainrun updated (running times, updated path)
    - [ ] Create invalid trainrun => train schedule creation (simulation not working, rolling stock must be entered in OSRD first) => once the rollingstock is given and simulation failed => trainrun deleted / updated (TBD)
  - [ ] Update trainrun
    - [x] Update trainrun path (shorten and lengthen steps or only start, intermediary or end step) => train schedule updated and simulation done => trainrun updated (updated running times, updated path)
    - [ ] Update trainrun departure time => train schedule updated and simulation done => trainrun updated (updated running times, updated path)
    - [ ] Update trainrun label => create/update/delete label of associated train schedule
  - [x] Delete trainrun
    - [x] Delete trainrun => delete train schedule

*a `trainrun` in NGE is the equivalent of a `train schedule` in OSRD (even tho it is not semantically 100% correct)

## Implementation plan

- [ ] [OSRD <- NGE] Implement events handler for NGE operations (in OSRD front) => requests to update timetable
  - [x] Enable [`eventListener`](https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener) for the NGE `operationX` events
  - [ ] Callback functions to update back-end (timetable)
  - [ ] Unitary and functionnal tests

## Description

![OSRD-NGE Concept diagram](https://github.com/OpenRailAssociation/osrd/assets/57864277/fc52939b-c02a-4475-933f-c7f8ca62ebea)

### Train schedules v2

Train schedules v2 is not released at the time on the refinement of this ticket, refinement taking in account TSv2 will have to be done during the development of this ticket.

### [POSTPONE] Train frequency

Train frequency is not a concept well used in OSRD as well as it is in NGE, however this is a key concept to understand how to work with these two level of detail.

A train frequency helper will have to be implemented on OSRD side, a service that will "translate" a trainrun in NGE into a serie of train schedules in OSRD.

For example, if a trainrun, with a trainrun frequency of 1 hour, is created, it will result in the creation of X (to be defined, maybe 24) train schedules in OSRD, with a very close name (i.e., "{trainrun.name}-{hourOfDeparture}), all shifted from 1 hour, using the same rolling stock, same minute of departure, same path etc...

For you information, in the other hand, when converting a scenario from OSRD to NGE, when a serie of train schedules that have the same attributes enough to be interpreted as frequenced train schedules (same rolling stock, slight variation of name (i.e., "{trainrun.name}-{hourOfDeparture}), same path, evenly shifted in time (every hour / 2 hours...)), it will have to let to display only a single trainrun in NGE (frequency depending on the frequence pattern observed) 

### Payloads

See [Trainrun data model](https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-frontend/blob/main/src/app/models/trainrun.model.ts) and [TrainrunSection data model](https://github.com/SchweizerischeBundesbahnen/netzgrafik-editor-frontend/blob/main/src/app/models/trainrunsection.model.ts)

**Create** `trainrun`:
```javascript
TrainrunSectionOperation {
    "type": "create",
    "trainrunSection": TrainrunSection
}
```

**Update** `trainrun`:
```javascript
TrainrunOperation {
    "type": "update",
    "trainrun": Trainrun
}
```

**Delete** `trainrun`:
```javascript
TrainrunOperation {
    "type": "delete",
    "trainrun": Trainrun
}
```

### Timetable update

The philosophy behind the timetable update is different from the PoC, since in the PoC the whole NGE model is computed and the result is compared to the whole timetable. Here we aim to deal with unitary action from NGE and request for a timetable update only once (atomic change instead of whole timetable comparison). In addition to that, the PoC has been implemented on TSv1. For these reasons, do not stick to the PoC algorithms (even tho it can help to understand).

### Labels

The creation or update of a label of a trainrun has to be persisted in OSRD train schedule.

To start with, trainrun labels can be set as train schedule labels, such as:
- `category` ("IC", "EC", ...)
- `frequency` (15min, 30min, ...)

## Definition of ready

- **PO**
  - [x] Acceptance criteria are clear and understandable

- **Technical**
  - [x] If required documentation is available on osrd.fr
  - [x] Implementation plan is detailed and validated by maintainers

- **General**
  - [x] WSJF values are filled
  - [x] Validated by the PM

**Comment by maelysLeratRosso** [comment]
# CR atelier 29/07/24
**Présents : Louis, Uriel, Simon, Florian, Maëlys**

### Questions 
- Quand on crée un train dans NGE, dans l'attribut `path` de `train_schedule`, que mettre par défaut pour l'attribut `deleted `? -> `false `(à la création depuis NGE, mais ne pas écraser depuis NGE s'il l'attribut est à `true`)
- Quand on crée un train dans NGE, dans l'attribut `schedule` de `train_schedule `, que mettre par défaut pour les attributs `locked `et `on_stop_signal `? -> `false` pour les 2 (à la création depuis NGE, mais ne pas écraser depuis NGE s'il l'attribut est à `true`)
- Quand on crée un train dans NGE, que mettre par défaut dans l'attribut `rolling_stock `de train_schedule? `""`  (à la création depuis NGE, mais ne pas écraser depuis NGE s'il un autre rolling stock est renseigné dans la circulation)
- Comment garder en mémoire dans OSRD les attributs de `TrainCategory` et de `Frequency` inscrits dans NGE ? 
  - -> **Court terme** : 
    - on place ces attributs dans des labels, auxquels un préfixe est associé pour reconnaître dans NGE quels labels utiliser (respectivement `category::` et` frequency::`) sur la base de [scoped-labels de GitLab](https://docs.gitlab.com/ee/user/project/labels.html#scoped-labels). 
    - Si aucun label n'est présent sur une circulation, utiliser une valeur par défaut dans NGE (**TODO** lister attributs par défaut pour les 2)
    - Si 2 labels sont présents sur une circulation pour le même préfixe, utiliser une valeur par défaut dans NGE (**TODO** comment discriminer ? remplacer l'ancien label par le nouveau ?)
    - La liste de labels présentés dans NGE à la création/édition d'un train est : UNION(une liste par défaut + ceux des circulations de la grille pour le préfixe associé) (**TODO** liste par défaut + exemple ceux de la circu)
  - -> **Moyen terme** : forcer l'unicité des labels pour chaque préfixe dans OSRD? (**TODO** remplacer l'ancien ou liste de priorité?)
  - -> **Long terme** : ces informations seront sans doute gardées dans la famille de train (au moins pour la fréquence)
- Comment garder en mémoire dans OSRD les attributs de `numberOfStrops `inscrits dans NGE? -> Pour le moment, on ne gère pas cet attribut.

### TODO sur ce PI 
- [ ] Supporter dans NGE le fait d'avoir une `TrainrunSection` sans horaires/temps de parcours, tout en permettant d'ajouter des valeurs (l'heure de départ de la TrainRun est obligatoire)
- [ ] Les labels associés à la catégorie de train et à la fréquence sont transmis à NGE
- [ ] Le chemin demandé (`path`) est transmis à NGE (seulement l'input)
- [ ] Les points horairisés (`schedule`) sont transmis à NGE sur les 4 points horairisés de NGE
- [ ] Ajouter la notion de CH dans les noeuds NGE pour que deux PRs avec le même trigramme ne soient pas mergés en un noeud NGE (exemple : DN/BV et DN/SG doivent être deux noeuds distincts) -> Concaténer `{$trigramm/$ch}` dans le BP (= Betriebspunktname = shortname = trigramme lol) de NGE
- [ ] Afficher une erreur dans NGE si un noeud par lequel passe une circulation est renommé ([duplicat ?](https://github.com/osrd-project/osrd-confidential/issues/593)
- [ ] Ne pas clean les noeuds non empruntés par une circulation (on initialise l'affichage des noeuds NGE par les noeuds empruntés par les circulations, mais à la synchronisation, on ne supprime pas) ([duplicat ?](https://github.com/osrd-project/osrd-confidential/issues/593)
- [ ] S'il reste du temps, conserver les noeuds NGE dans localStorage (cf critères d'acceptance https://github.com/osrd-project/osrd-confidential/issues/593 et https://github.com/osrd-project/osrd-confidential/issues/597)

### TODO côté métier 
- Vérifier comment gérer les temps d'arrêt qui sont renseignés modulo 60 (voir si on peut différencier dans NGE un arrêt de 10 minutes d'un arrêt de 70 minutes)
